### PR TITLE
module: Add thermal support

### PR DESCRIPTION
**Pull request by punitagrawal**
Add a thermal module that will allow querying for thermal zones and trip
points in the target. The module allow supports enable/disabling thermal
zones as well as change trip temperatures.

Signed-off-by: Punit Agrawal <punit.agrawal@arm.com>

### CODE REVIEW

File: devlib/module/thermal.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    @staticmethod
+    def probe(target):
+
+        if target.file_exists(ThermalModule.thermal_root):
+            return True
+
+    def __init__(self, target):
+        super(ThermalModule, self).__init__(target)
+
+        self.logger = logging.getLogger('Thermal')
```
Comment (setrofim) [inline devlib/module/thermal.py]: this is unnecessary -- `self.logger` is initialized in the super

File: devlib/module/thermal.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+        for entry in target.list_directory(self.thermal_root):
+            re_match = re.match('^(thermal_zone|cooling_device)([0-9]+)', entry)
+
+            if re_match.group(1) == 'thermal_zone':
+                self.thermal_zone(re_match.group(2))
+            elif re_match.group(1) == 'cooling_device':
+                pass
+
+    def thermal_zone(self, _id):
```
Comment (setrofim) [inline devlib/module/thermal.py]: should be `add_thermal_zone`

File: devlib/module/thermal.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        self.zones = {}
+        self.cdevs = []
+
+        for entry in target.list_directory(self.thermal_root):
+            re_match = re.match('^(thermal_zone|cooling_device)([0-9]+)', entry)
+
+            if re_match.group(1) == 'thermal_zone':
+                self.thermal_zone(re_match.group(2))
+            elif re_match.group(1) == 'cooling_device':
+                pass
```
Comment (setrofim) [inline devlib/module/thermal.py]: Why match it and then ignore it? Is this a `TODO`?

File: devlib/module/thermal.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        value = 'disable'
+        if enable:
+            value = 'enabled'
+
+        self.target.write_value(self.target.path.join(self.path, 'mode', value))
+
+    def get_temperature(self):
+        """Returns the temperature of the thermal zone"""
+        temp_file = self.target.path.join(self.path, 'temp')
+        return self.target.read_value(temp_file)
```
Comment (setrofim) [inline devlib/module/thermal.py]: `self.target.read_int(temp_file)`?

File: devlib/module/thermal.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+            return True
+        else:
+            return False
+
+    def set_mode(self, enable):
+        value = 'disable'
+        if enable:
+            value = 'enabled'
+
+        self.target.write_value(self.target.path.join(self.path, 'mode', value))
```
Comment (setrofim) [inline devlib/module/thermal.py]: Closing paren too far to the right. Should be

    self.target.write_value(self.target.path.join(self.path, 'mode'), value)

File: devlib/module/thermal.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        if value == 'enabled':
+            return True
+        else:
+            return False
+
+    def set_mode(self, enable):
+        value = 'disable'
+        if enable:
+            value = 'enabled'
+
```
Comment (setrofim) [inline devlib/module/thermal.py]: above could be 

    value = 'enabled' if enable else 'disabled'

File: devlib/module/thermal.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    def trip_point(self, _id):
+        self.trip_points[int(_id)] = TripPoint(self, _id)
+
+    def is_enabled(self):
+        """Returns a boolean representing the 'mode' of the thermal zone"""
+        value = self.target.read_value(self.target.path.join(self.path, 'mode'))
+        if value == 'enabled':
+            return True
+        else:
+            return False
```
Comment (setrofim) [inline devlib/module/thermal.py]: Could be just

      return value == 'enabled'

File: devlib/module/thermal.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        self.name = 'thermal_zone' + _id
+        self.path = target.path.join(root, self.name)
+        self.trip_points = {}
+
+        for entry in self.target.list_directory(self.path):
+            re_match = re.match('^trip_point_([0-9]+)_temp', entry)
+            if re_match is not None:
+                self.trip_point(re_match.group(1))
+
+    def trip_point(self, _id):
```
Comment (setrofim) [inline devlib/module/thermal.py]: should be `add_trip_point`

File: devlib/module/thermal.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        self.type_node = 'trip_point_' + _id + '_type'
+
+    def get_temperature(self):
+        """Returns the currently configured temperature of the trip point"""
+        temp_file = self.zone.target.path.join(self.zone.path, self.temp_node)
+        return self.zone.target.read_value(temp_file)
+
+    def set_temperature(self, temperature):
+        temp_file = self.zone.target.path.join(self.zone.path, self.temp_node)
+        self.zone.target_write_value(temp_file, temperature)
```
Comment (setrofim) [inline devlib/module/thermal.py]: I assume this should be

    self.zone.target.write_value(temp_file, temperature)

`self.zone.target` gets used a lot here; might be cleaner to expose it as `self.target` via a property:

    @property:
    def target(self):
         return self.zone.target

File: devlib/module/thermal.py (hunk 1 above)
Comment (punitagrawal) [inline devlib/module/thermal.py]: Dropped the logger initialisation here.

File: devlib/module/thermal.py (hunk 2 above)
Comment (punitagrawal) [inline devlib/module/thermal.py]: Changed the name

File: devlib/module/thermal.py (hunk 3 above)
Comment (punitagrawal) [inline devlib/module/thermal.py]: Indeed, it's a TODO. I have implemented what I needed - hopefully over time we can cover most thermal related operations in this module.

File: devlib/module/thermal.py (hunk 4 above)
Comment (punitagrawal) [inline devlib/module/thermal.py]: I hadn't discovered read_int yet. Now updated to use it.

File: devlib/module/thermal.py (hunk 5 above)
Comment (punitagrawal) [inline devlib/module/thermal.py]: Oops! Thanks for spotting - not sure what my test was doing if this is broken.

File: devlib/module/thermal.py (hunk 6 above)
Comment (punitagrawal) [inline devlib/module/thermal.py]: Nice! I am showing my lack of python chops here.

File: devlib/module/thermal.py (hunk 7 above)
Comment (punitagrawal) [inline devlib/module/thermal.py]: Done.

File: devlib/module/thermal.py (hunk 9 above)
Comment (punitagrawal) [inline devlib/module/thermal.py]: Fixed.

I've also made target a property. Perhaps, in the ThermalZone class as well I should refer to the module target. Hmm...

File: devlib/module/thermal.py (hunk 3 above)
Comment (setrofim) [inline devlib/module/thermal.py]: Cool; could you add a comment tagged with "TODO" just so it's clear.

**Pull request merged**